### PR TITLE
Improve codegen for world validation

### PR DESCRIPTION
**Pull request by JoJoJet**
# Objective

Improve code-gen for `QueryState::validate_world` and `SystemState::validate_world`.

## Solution

* Move panics into separate, non-inlined functions, to reduce the code size of the outer methods.
* Mark the panicking functions with `#[cold]` to help the compiler optimize for the happy path.
* Mark the functions with `#[track_caller]` to make debugging easier.

### CODE REVIEW

**Review by mockersf — APPROVED** [review]
I think I remember this being better, and it's a pattern already used in Bevy, but don't remember if it has any actual impact

File: crates/bevy_ecs/src/query/state.rs [hunk 1]
```diff
@@ -189,12 +189,17 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
...
-            world_id == self.world_id,
-            "Attempted to use {} with a mismatched World. QueryStates can only be used with the World they were created from.",
-                std::any::type_name::<Self>(),
-        );
+        #[inline(never)]
+        #[track_caller]
+        #[cold]
+        fn panic_mismatched(this: WorldId, other: WorldId) -> ! {
+            panic!("Encountered a mismatched World. This QueryState was created from {this:?}, but a method was called using {other:?}.");
+        }
```
Comment (james7132) [inline crates/bevy_ecs/src/query/state.rs:199]:
```suggestion
        }
        
```

File: crates/bevy_ecs/src/system/function_system.rs [hunk 2]
```diff
@@ -216,8 +216,17 @@ impl<Param: SystemParam> SystemState<Param> {
...
     #[inline]
+    #[track_caller]
     fn validate_world(&self, world_id: WorldId) {
-        assert!(self.matches_world(world_id), "Encountered a mismatched World. A SystemState cannot be used with Worlds other than the one it was created with.");
+        #[inline(never)]
+        #[track_caller]
+        #[cold]
+        fn panic_mismatched(this: WorldId, other: WorldId) -> ! {
+            panic!("Encountered a mismatched World. This SystemState was created from {this:?}, but a method was called using {other:?}.");
+        }
```
Comment (james7132) [inline crates/bevy_ecs/src/system/function_system.rs:226]:
```suggestion
        }
        
```

**Review by james7132 — COMMENTED** [review]
Small style nits. The motivation and logic behind the change makes sense to me. Could you rebase this PR to the latest main? I want to test this against bevy_asm_tests to see the output difference.

**Review by james7132 — APPROVED** [review]
[It does seem to work](https://github.com/james7132/bevy_asm_tests/commit/001630a7f92bce35d5ee2e1cff9252ac71ff7f74#diff-496530101f0b16e495b7e9b77c0e906ae3068c8adb69ed36c92d5a1be5a9efbeR899). There was a rough 300 fewer lines of assembly generated across several test cases for QueryState, a 14% decrease in codegen. Though the bulk of that seems to be in the symbols for panicking, this is still a decent win.

**Pull request merged**